index.js

```diff
@@ -61,17 +77,22 @@ app.use(async (ctx) => {
     ctx.status = 405;
     return;
   }
-
+  
   if (ctx.ws) {
     const ws = new WebSocketJSONStream(await ctx.ws());
-    db.listen(ws, docId);
+    db.listen(ws, {docId, readOnly}); // docId and readOnly is passed to 'connect' middleware as ctx.req
   } else {
-    ctx.body = 'Document exists.';
+    ctx.body = {docId, readOnly};
   }
 });
 
 app.listen(process.env.PORT || 8080);
 
-function isEmptyObject(obj) {
-  return Object.keys(obj).length === 0 && obj.constructor === Object;
+function getSessionDetails(sessionId) {
+  for (const session of documents) {
+    if (session.has(sessionId)) {
+      return [session.get("docId"), session.get(sessionId)];
+    }
+  }
```

Comment:
Is there a faster way to check for membership than performing an `O(n)` iteration on potentially an unbounded number of documents?

Reply:
Maybe instead of maintaining a set, I could just maintain a map of (key,value) pairs. The keys will be the short ids since that's the only information the backend receives when a client connects to the session, and the values will be an object/array of docId and readOnly status. This will change lookup time to just O(1)